src/main/scala/resource/SharedResource.scala

```diff
@@ -0,0 +1,59 @@
+package resource
+
+import java.util.concurrent.atomic.AtomicReference
+import java.util.concurrent.atomic.AtomicInteger
+
+final class SharedResource[A : Resource : Manifest](opener: => A) extends AbstractManagedResource[A] {
+	private val resource = implicitly[Resource[A]]
+	// A referencing counting atomic variable we use to "guard" this resource.
+	// Currently, we spin-lock accessing this ref because we don't expect opening to block a thread for very long.
+	private val ref = new AtomicReference[Option[A]](None)
+	private val count = new AtomicInteger(0)
+
+	protected def open: A = {
+		val myCount = count.addAndGet(1)
+		// If we win the race, we construct the resource
+		if (myCount == 1) {
+			// We must construct the reference
+			// We may want to CAS here and fail if we're inconsistent.
+			val r = opener
```

Comment:
Good point.  That will literally cause all threads waiting for the resource to be blocked.  This confirms my suspicion we need some kind of an error channel to signal to all threads that the resource has failed and they need to bail.   Let me think about how to handle this appropriately....